services/tns_submission_queue/tns_submission_queue.py

```diff
@@ -52,6 +52,618 @@ class TNSReportError(Exception):
+        The database session to use.
+    """
+    # if the bot is set up to only report objects to TNS if they are not already there,
+    # we check if an object is already on TNS (within 2 arcsec of the object's position)
+    # and if it is, we skip the submission
+    # otherwise, we submit as long as there are no reports with the same internal source name
+    # (i.e. the same obj_id from the same survey)
+    altdata = tnsrobot.altdata
+    obj_id = submission_request.obj_id
+
+    _, existing_tns_name = get_IAUname(altdata['api_key'], tns_headers, obj_id=obj_id)
+    if existing_tns_name is not None:
+        if not tnsrobot.report_existing:
+            raise TNSReportError(
+                f'{obj_id} already posted to TNS as {existing_tns_name}.'
+            )
+        else:
+            # look if the object on TNS has already been reported by the same survey (same internal name, here being the obj_id)
+            internal_names = get_internal_names(
+                altdata['api_key'], tns_headers, tns_name=existing_tns_name
+            )
+            if len(internal_names) > 0 and obj_id in internal_names:
+                raise TNSReportError(
+                    f'{obj_id} already posted to TNS with the same internal source name.'
+                )
+
+
+def find_source_to_submit(submission_request, tnsrobot_groups, session):
+    """Find the source to submit to TNS.
+
+    Parameters
+    ----------
+    submission_request : `~skyportal.models.TNSRobotSubmission`
+        The submission request.
+    tnsrobot_groups : list of `~skyportal.models.TNSRobotGroup`
+        The TNSRobotGroups that the user is allowed to submit to with this robot.
+    session : `~sqlalchemy.orm.Session`
+        The database session to use.
+
+    Returns
+    -------
+    source : `~skyportal.models.Source`
+        The source to submit.
+    """
+    # Get the sources saved to the groups that have access to this TNS robot,
+    # this is so if a user that has access to this robot saved the obj as a source before,
+    # he is the first author when another user auto-submits the source
+    tnsrobot_id = submission_request.tnsrobot_id
+    obj_id = submission_request.obj_id
+    source = session.scalar(
+        sa.select(Source)
+        .where(
+            Source.obj_id == obj_id,
+            Source.active.is_(True),
+            Source.group_id.in_(
+                [tnsrobot_group.group_id for tnsrobot_group in tnsrobot_groups]
+            ),
+        )
+        .order_by(Source.saved_at.asc())
+    )
+
+    if source is None:
+        if submission_request.auto_submission:
+            raise TNSReportError(
+                f'No source {obj_id} saved to any group with TNSRobot {tnsrobot_id}.'
+            )
+        else:
+            raise TNSReportError(
+                f'No source {obj_id} saved to group with TNSRobot {tnsrobot_id}.'
+            )
+
+
+def build_reporters_string(submission_request, source, tnsrobot, session):
```

Comment:
This is the type of logic that is quite easy to unit test outside of the web framework.